### PR TITLE
Feed forward layer, frontend and encoder

### PR DESCRIPTION
**Pull request by DanEnergetics**
Implementation of a simple feed forward encoder that serves to generate good alignments in full-sum HMM training. Apart from the layers, it implements a simple convolutional front-end that acts as a feed-forward layer that takes a window of feature time frames into account.

### CODE REVIEW

File: i6_models/parts/frontend/window_ffnn.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    def forward(self, x: torch.Tensor, /, sequence_mask: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        T might be reduced to T' on stride
+
+        :param x: input tensor of shape [B,T,F]
+        :param sequence_mask: the sequence mask for the tensor
+        :return: torch.Tensor of shape [B,T',F'] and the shape of the sequence mask
+        """
+        x = x.transpose(1, 2)  # torch 1d convolution is over last dim but we want time conv
+        x = self.conv(x).transpose(1, 2)
```
Comment (michelwi) [inline i6_models/parts/frontend/window_ffnn.py:76]: I think we usually add annotations about the axes in comments, especially if we change them
```suggestion
        # torch 1d convolution is over last dim but we want time conv
        x = x.transpose(1, 2)  # [B, F, T]
        x = self.conv(x).transpose(1, 2)  # [B, T', F']
```

File: i6_models/parts/frontend/window_ffnn.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+        # these settings apparently apply stride correctly to the masking whatever the kernel size
```
Comment (michelwi) [inline i6_models/parts/frontend/window_ffnn.py:78]:
```suggestion
        # change masking according to stride value
```
do be confident about the implementation of others xP

File: i6_models/parts/frontend/window_ffnn.py
Comment (michelwi) [inline i6_models/parts/frontend/window_ffnn.py:1]: I would maybe not name it "ffnn" / "FeedForward" but rather "convolution" since that is the layer that is being used?

Comment (DanEnergetics) [reply i6_models/parts/frontend/window_ffnn.py:1]: I was unsure about naming and open to changes. Would you then rename both file and classes? Something like `window_convolution.py` then for the file?

File: i6_models/parts/frontend/window_ffnn.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    in_features: int
+    out_features: int
+    dropout: float
+    window_size: int
+    stride: int
+    activation: Union[nn.Module, Callable[[torch.Tensor], torch.Tensor]]
+
+    def __post_init__(self):
+        super().__post_init__()
+        assert self.window_size % 2 == 1, "Only odd kernel sizes are supported so far"
```
[the remark this answers is not in the document]
Comment (DanEnergetics) [reply i6_models/parts/frontend/window_ffnn.py:39]: I actually found the same issue as mentioned in https://github.com/rwth-i6/i6_models/issues/50. It still does not work with even padding. However, I have a PR lined up with a simple padding function that would fix this.

File: i6_models/parts/frontend/window_ffnn.py (hunk 2 above)
Comment (DanEnergetics) [reply i6_models/parts/frontend/window_ffnn.py]: Actually I was not confident in my own implementation here. I had to choose `kernel_size = 1` and `padding = 0` to achieve correct masking, which is not how this function is supposed to be used I think.

File: i6_models/parts/frontend/window_ffnn.py
Comment (michelwi) [reply i6_models/parts/frontend/window_ffnn.py:1]: yes, sounds good to me

File: i6_models/parts/ffnn.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import torch.nn.functional as F
+
+from i6_models.config import ModelConfiguration
+
+
+@dataclass
+class FeedForwardLayerV1Config(ModelConfiguration):
+    """
+    Attributes:
+        in_features: input feature dimension
```
[the remark this answers is not in the document]
Comment (michelwi) [reply i6_models/parts/ffnn.py:18]: I think I would have preferred the parameter naming `input_dim` + `output_dim` (or use the same names as in `torch.nn.Linear`) but now it is at least consistent. :)

File: i6_models/parts/frontend/window_ffnn.py (hunk 2 above)
Comment (michelwi) [reply i6_models/parts/frontend/window_ffnn.py]: And with `kernel_size = cfg.window_size` and `padding = get_same_padding(cfg.window_size)` (i.e. the values given to `Conv1d`) it does not work?

**Review by michelwi — APPROVED** [review]

File: i6_models/parts/frontend/window_ffnn.py (hunk 2 above)
Comment (DanEnergetics) [reply i6_models/parts/frontend/window_ffnn.py]: Weirdly enough yes. I came up with a failing test case in a new branch https://github.com/rwth-i6/i6_models/blob/55a6fd6542037ed437a2a8d6d85ad98a1e8d73ee/tests/test_masking.py#L7

**Review by Atticus1806 — APPROVED** [review]
Took the liberty to commit one last change :)
